src/main/java/com/amazon/ion/bytecode/util/BytecodeBuffer.kt

```diff
@@ -0,0 +1,269 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+package com.amazon.ion.bytecode.util
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings
+
+/**
+ * This is a custom collection that allows unsafe access to the backing array for storing bytecode instructions.
+ *
+ * It provides specialized methods for adding single values, pairs, triples, and slices of bytecode instructions
+ * efficiently. The buffer automatically grows as needed using a growth multiplier strategy.
+ *
+ * Potential Performance Improvement: Consider exposing raw, `@JvmField`-annotated fields for `size` and `capacity` if it will improve the performance.
+ */
+internal class BytecodeBuffer private constructor(
```

Comment:
This is super similar to ConstantPool... but I'm sure you've already noticed that and decided we need both due to issues with adding another layer of abstraction.